DSDEGP-1275 stub in an /authorization endpoint.

tbl3rd opened the pull request:
### Description

_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any JIRA issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [X] **Submitter**: Include the JIRA issue number in the PR description
- [X] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [X] **Submitter**: Ensure that you have added or updated tests
- [X] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

tbl3rd committed:
Accept and reject tests both work.

tbl3rd committed:
Factor methods into a trait.

tbl3rd committed:
Change value from def to val.

tbl3rd committed:
Back out debug println()s.

tbl3rd committed:
Rename auth to authorization consistently.

tbl3rd commented:
This PR does not authorize against all requests against a whitelist.
It just implements an authorizer to pass to the authorize() directive
and wraps a stub /authorization endpoint with it, then adds accept
and reject tests to validate the OIDC headers.

tbl3rd committed:
Export test /authorization endpoint (and consistency).

pshapiro4broad commented on Auth.org:
```diff
@@ -0,0 +1,66 @@
+#+TITLE: Clio Authentication and Authorization
```
What is this file? It looks like documentation but if so I'm not sure why it's called `Auth.org` and is stored at the root of the clio repo.

tbl3rd replied:
It just documents my current understanding.
It will wind up in the README, Confluence,
or deleted eventually.

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/model/authorization.scala:
```diff
@@ -0,0 +1,65 @@
+import scala.util.Try
+
+
+trait OidcHeader {
```
I'm not sure what our conventions are in Scala, but I didn't expect to see a class named `OidcHeader` in a file called `authorization.scala`. I don't know why `Oidc` is; is that something specific to this implementation or a general concept about authorization?

There should be comments here too.

kshakir replied:
Re the naming _only_: from the [clio wiki](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509#Clio(MetadataManager)-Filenaming) and borrowed from the [scala style guide](http://docs.scala-lang.org/style/files.html):

> As a rule, files should contain a single logical compilation unit. … One exception to this guideline is for classes or traits which have companion objects. … These files should be named according to the class, trait or object they contain …
> All multi-unit files should be given camelCase names with a lower-case first letter. This is a very important convention. It differentiates multi- from single-unit files, greatly easing the process of finding declarations.

tbl3rd replied:
It used to have more authy named things in it.
So I should rename it 'oidcHeaders.scala'?

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/service/AuthorizationService.scala:
```diff
@@ -0,0 +1,34 @@
+  def authorize(context: RequestContext): Boolean = {
+    val pairs = for (header <- context.request.headers) yield
+      header match {
+        case OidcAccessToken(token)      => 0 -> token
```
It's not clear to me what the numbers here represent. Is it just a way to allow the code to allow the API to provide either `user id` or `sub` in the header? Is this a common idiom in Akka code?

IMO it might be clearer to have the code make a map using strings instead of numbers, e.g. `"token" -> token` and then you don't need `Vector` below to unpack them, you can pull them out directly using a map lookup.

tbl3rd replied:
The numbers just represent parameters to AuthorizationInfo.
I hope there is a solution better than yours (and mine).

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/service/AuthorizationService.scala:
```diff
@@ -0,0 +1,34 @@
+      }
+    (pairs.size == 4) && (Vector.tabulate(4) { pairs.toMap } match {
+      case Vector(token, expires, email, id) =>
+        AuthorizationService.mock == AuthorizationInfo(token, expires, email, id)
```
I'm unclear what `.mock` is used for here. I would expect it to be a mocked service API for testing but it looks like it contains the actual values from the HTTP response header.

It looks like this is changing the value in `AuthorizationService.mock` but I don't see how that could work since it's declared as a `val`.

tbl3rd replied:
We don't have an authentication service yet.
This is really just plumbing now.

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/ClioServer.scala:
```diff
@@ -32,7 +33,7 @@ object ClioServer
     auditRequest & auditResult & completeWithInternalErrorJson & auditException & mapRejectionsToJson
   }
-  private val innerRoutes: Route = concat(statusRoutes)
+  private val innerRoutes: Route = concat(authorizationRoute, statusRoutes)
```
This adds `/authorization` as new endpoint but it doesn't make authorization required by any other APIs, correct?

tbl3rd replied:
Correct.

tbl3rd replied:
The /authorization route is there for testing.
It isn't even exported as an endpoint yet --
though more by default than by design.

pshapiro4broad commented on src/test/scala/org/broadinstitute/clio/webservice/AuthorizationWebServiceSpec.scala:
```diff
@@ -0,0 +1,67 @@
+import org.broadinstitute.clio.model._
+import org.broadinstitute.clio.service._
+
+class AuthWebServiceSpec extends FlatSpec
```
Should the code also test the case where both `sub` and `user id` are provided, since it supports that case?

tbl3rd replied:
I don't know.  Maybe.

pshapiro4broad commented on src/test/scala/org/broadinstitute/clio/webservice/AuthorizationWebServiceSpec.scala:
```diff
@@ -0,0 +1,67 @@
+  it should "authorizationRoute" in {
+
+    Get("/authorization").
+      addHeader(OidcClaimExpiresIn(AuthorizationService.mock.expiration)).
```
Why is this code using `AuthorizationService.mock` to provide the value for `expiration`? It looks like any value would work here.

tbl3rd replied:
Any value would work here now.
But "AuthWebService" should be "AuthorizationWebService".

pshapiro4broad reviewed (approved):
Many comments but seems OK overall

👍

tbl3rd commented on src/test/scala/org/broadinstitute/clio/webservice/AuthorizationWebServiceSpec.scala:
```diff
@@ -0,0 +1,67 @@
+    authorizationRoute ~> check {
+      status shouldEqual StatusCodes.OK
+    }
+  }
```
It seems to me that the AuthorizationService is more like the AuditService
than like the StatusService, and it should probably be tested similarly.

tbl3rd committed:
Route tests succeed again.

tbl3rd committed:
Implement basic integration test for /authorization.

Respond to some review comments.  More later.

tbl3rd committed:
Mock OIDC headers with their names.

Why is response 403:Forbidden instead of 401:Unauthorized?

tbl3rd committed:
Move header classes into web service ...

... to be renamed directives somehow.

tbl3rd committed:
AuthorizationWebService -=> AuthorizationDirectives

tbl3rd committed:
Clarify test messages.

tbl3rd committed:
Move AuthorizationDirectives to the top.

tbl3rd committed:
Move headers to new file.

tbl3rd committed:
Remove debug cruft.

tbl3rd committed:
Rename OidcHeader -=> RequestResponseHeader.

tbl3rd committed:
Factor out and document OIDC classes.

Also tried to use the Akka HTTP header parsing machinery with
OidcClaimExpiresIn, but pretty sure I don't like the result.
